Multiple tasks in the same job that generate SARIF output collide due to job id being used

**Describe the bug**
When a pipeline job has multiple tasks which both write the same kind of additional output, there is a collision.

Additionally, when the tasks are rendered, selecting each task offers the additional output for any task in from the job, report download needs to have its scope tightened to the task.

**Trivy Azure Pipeline Task Verison**
1.12.0

**Hosted or Self-hosted Agent**
Both affected

**Trivy as Docker?**
Both affected

**Error message**
```
##[error]Artifact 5ff6b4dc-ffc7-54d9-7597-f9befd3c7152SARIF already exists for build 259.
```

https://dev.azure.com/owenrumney0307/trivy-dev/_build/results?buildId=259&view=logs&j=5ff6b4dc-ffc7-54d9-7597-f9befd3c7152&t=30f6d104-a5ca-5c2c-8248-64f1bfdb0c95&l=43

--- a/.azuredevops/pipelines/integration.yaml
+++ b/.azuredevops/pipelines/integration.yaml
@@ -23,7 +23,13 @@ stages:
               docker: false
               version: $(trivyVersion)
               image: ubuntu:latest
+              cosignOutput: true
+              cyclonedxOutput: true
+              jsonOutput: true
               sarifOutput: true
+              spdxjsonOutput: true
+              spdxOutput: true
+              tableOutput: true
               exitCode: 0
 
           - task: trivy-dev@1
@@ -33,6 +39,12 @@ stages:
               trivyImage: ghcr.io/aquasecurity/trivy:latest
               image: ghcr.io/aquasecurity/trivy:latest
               cosignOutput: true
+              cyclonedxOutput: true
+              jsonOutput: true
+              sarifOutput: true
+              spdxjsonOutput: true
+              spdxOutput: true
+              tableOutput: true
               exitCode: 0
 
           - bash: ls -lR "${AGENT_TEMPDIRECTORY}/trivy"

--- a/.azuredevops/pipelines/integration.yaml
+++ b/.azuredevops/pipelines/integration.yaml
@@ -25,9 +25,7 @@ stages:
               image: ubuntu:latest
               cosignOutput: true
               cyclonedxOutput: true
-              jsonOutput: true
               sarifOutput: true
-              spdxjsonOutput: true
               spdxOutput: true
               tableOutput: true
               exitCode: 0
@@ -40,11 +38,8 @@ stages:
               image: ghcr.io/aquasecurity/trivy:latest
               cosignOutput: true
               cyclonedxOutput: true
-              jsonOutput: true
               sarifOutput: true
               spdxjsonOutput: true
-              spdxOutput: true
-              tableOutput: true
               exitCode: 0
 
           - bash: ls -lR "${AGENT_TEMPDIRECTORY}/trivy"

--- a/trivy-task/additionalReporting.ts
+++ b/trivy-task/additionalReporting.ts
@@ -1,6 +1,6 @@
+import path from 'path';
 import task = require('azure-pipelines-task-lib/task');
 import { createRunner } from './trivyLoader';
-import { randomSuffix } from './utils';
 
 const reportTypes = {
   sarif: { DisplayName: 'SARIF', Extension: '.json' },
@@ -13,43 +13,31 @@ const reportTypes = {
 };
 
 export async function generateAdditionalReports(filename: string) {
-  const jobId = task.getVariable('System.JobId') || '';
-  const smallJobId = jobId.substring(0, 8);
+  const artifactName = path.basename(filename, path.extname(filename));
 
   for (const [key, value] of Object.entries(reportTypes)) {
     task.debug(`Checking if ${key} report is required`);
     const inputKey = `${key}Output`;
     const outputKey = `${key}Report`;
+
     if (task.getBoolInput(inputKey, false)) {
       if (key === 'json') {
-        // don't need to convert json to json
+        // don't need to convert json to json and create attachment
         task.setVariable(outputKey, filename, false, true);
-        task.debug(`Uploading ${key} report...`);
-        const artifactKey = `${key}-${smallJobId}-${randomSuffix(8)}`;
-        task.uploadArtifact(
-          artifactKey,
-          filename,
-          `${jobId}${value.DisplayName}`
-        );
+        task.uploadArtifact(artifactName, filename, artifactName);
         continue;
       }
 
       console.log(`Generating ${key} report...`);
       const format = getReportFormat(key);
       const output = `${filename.replace(/json$/, format)}${value.Extension}`;
-      const localOutput = `${filename.replace(/json$/, format)}${value.Extension}`;
 
       try {
         await generateReport(format, output, filename);
         console.log(`Generated ${key} report at ${output}`);
-        task.setVariable(outputKey, localOutput, false, true);
-        task.debug(`Uploading ${key} report...`);
-        const artifactKey = `${key}-${smallJobId}-${randomSuffix(8)}`;
-        task.uploadArtifact(
-          artifactKey,
-          localOutput,
-          `${jobId}${value.DisplayName}`
-        );
+        task.addAttachment(outputKey, path.basename(output), output);
+        task.setVariable(outputKey, output, false, true);
+        task.uploadArtifact(artifactName, output, artifactName);
       } catch (error) {
         task.error(`Failed to generate ${key} report: ${error}`);
       }

--- a/trivy-task/index.ts
+++ b/trivy-task/index.ts
@@ -1,14 +1,15 @@
 import path from 'path';
 import task = require('azure-pipelines-task-lib/task');
 import { ToolRunner } from 'azure-pipelines-task-lib/toolrunner';
-
+import { randomUUID } from 'crypto';
 import { createRunner, tmpPath } from './trivyLoader';
 import { getAquaAccount, hasAquaAccount, isDevMode } from './utils';
 import { generateAdditionalReports } from './additionalReporting';
 
 async function run() {
   task.debug('Starting Trivy task...');
-  const resultsFileName = `results-${Math.random()}.json`;
+  const randomPrefix = randomUUID();
+  const resultsFileName = `trivy-results-${randomPrefix}.json`;
   const resultsFilePath = path.join(tmpPath, resultsFileName);
   task.rmRF(resultsFilePath);
 
@@ -44,7 +45,7 @@ async function run() {
   }
 
   const hasAccount = hasAquaAccount();
-  const assuranceFileName = `assurance-${Math.random()}.json`;
+  const assuranceFileName = `assurance-${randomPrefix}.json`;
   const assuranceFilePath = path.join(tmpPath, assuranceFileName);
 
   // copy the process env and add the aqua credentials

--- a/ui/src/App.tsx
+++ b/ui/src/App.tsx
@@ -111,12 +111,12 @@ export class App extends React.Component<AppProps, AppState> {
       setTimeout(this.check.bind(this), this.props.checkInterval);
       return;
     }
-    const attachments = await this.buildClient.getAttachments(
+    const jsonAttachments = await this.buildClient.getAttachments(
       this.project.id,
       build.id,
       'JSON_RESULT'
     );
-    if (attachments.length === 0) {
+    if (jsonAttachments.length === 0) {
       this.setState({
         error:
           'No attachments found: cannot load results. Did Trivy run properly?',
@@ -126,64 +126,88 @@ export class App extends React.Component<AppProps, AppState> {
 
     this.setState({ status: worstState });
 
-    const artifacts = await this.buildClient.getArtifacts(
-      this.project.id,
-      build.id
-    );
+    const reportTypes = {
+      sarifReport: 'SARIF',
+      cyclonedxReport: 'CycloneDX',
+      spdxReport: 'SPDX',
+      spdxjsonReport: 'SPDX JSON',
+      cosignReport: 'Cosign',
+      tableReport: 'Table',
+    };
+    type ReportType = keyof typeof reportTypes;
 
-    console.log(JSON.stringify(artifacts));
+    // get all supported report attachments for the build once
+    const additionalAttachments: Attachment[] = [];
+    for (const key of Object.keys(reportTypes)) {
+      const reportAttachments = await this.buildClient.getAttachments(
+        this.project.id,
+        build.id,
+        key
+      );
+      additionalAttachments.push(...reportAttachments);
+    }
 
-    attachments.forEach(
-      function (attachment: Attachment) {
-        records.forEach(
-          async function (record: TimelineRecord) {
-            try {
-              const buffer = await this.buildClient.getAttachment(
-                this.project.id,
-                build.id,
-                timeline.id,
-                record.id,
-                'JSON_RESULT',
-                attachment.name
-              );
-              const report = this.decodeReport(buffer) as Report;
-              if (!report.DownloadReports) {
-                report.DownloadReports = [];
-              }
+    jsonAttachments.forEach(
+      async function (attachment: Attachment) {
+        // get the record id from attachment url
+        const recordId = attachment._links.self.href.split('/')[10];
+        // get the record from the timeline
+        const record = records.find((record) => record.id === recordId);
+        if (!record) {
+          console.log(`Record not found for attachment: ${attachment.name}`);
+          return;
+        }
+        try {
+          const buffer = await this.buildClient.getAttachment(
+            this.project.id,
+            build.id,
+            timeline.id,
+            record.id,
+            'JSON_RESULT',
+            attachment.name
+          );
+          const report = this.decodeReport(buffer) as Report;
+          if (!report.DownloadReports) {
+            report.DownloadReports = [];
+          }
 
-              if (record.name) {
-                report.DisplayName = record.name;
-              }
+          if (record.name) {
+            report.DisplayName = record.name;
+          }
+          // Add json report to the report downloads by default
+          report.DownloadReports.push({
+            Name: 'JSON',
+            Url: attachment._links.self.href,
+          });
 
-              try {
-                console.log(
-                  'Checking for artifacts for record ' + JSON.stringify(record)
-                );
-                artifacts
-                  .filter((artifact) => artifact.source === record.parentId)
-                  .forEach((artifact) => {
-                    console.log('Artifact: ' + JSON.stringify(artifact));
-                    report.DownloadReports.push({
-                      // the name of the report has the task as a prefix so that needs
-                      // to be stripped
-                      Name: artifact.name.replace(artifact.source, ''),
-                      Url: artifact.resource.downloadUrl,
-                    });
-                  });
-              } catch {
-                console.log('Failed to decode report artifact');
-              }
-
-              this.setState((prevState: any) => ({
-                reports: [...prevState.reports, report],
-              }));
-            } catch (e) {
+          // check if there are any other attachments with the same record id
+          // and add them to the downloads
+          additionalAttachments
+            .filter(
+              (reportAttachment) =>
+                reportAttachment._links.self.href.split('/')[10] === recordId
+            )
+            .forEach((reportAttachment) => {
+              // get the report type from attachment url
+              const attachmentUrl = reportAttachment._links.self.href;
+              const attachmentType = attachmentUrl.split('/')[12];
               console.log(
-                'Failed to decode results attachment ' + JSON.stringify(e)
+                `Found ${attachmentType} for report ${report.DisplayName}`
               );
-            }
-          }.bind(this)
-        );
+              report.DownloadReports.push({
+                Name: reportTypes[attachmentType as ReportType],
+                Url: attachmentUrl,
+              });
+            });
+
+          this.setState((prevState: any) => ({
+            reports: [...prevState.reports, report],
+          }));
+        } catch (e) {
+          console.log(
+            'Failed to decode results attachment ' + JSON.stringify(e)
+          );
+        }
       }.bind(this)
     );
 

